File: src/main/java/de/gesellix/docker/client/filesocket/UnixSocket.java
package de.gesellix.docker.client.filesocket;

import org.newsclub.net.unix.AFUNIXSocket;
import org.newsclub.net.unix.AFUNIXSocketAddress;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.SocketException;

public class UnixSocket extends FileSocket {

    private static final Logger log = LoggerFactory.getLogger(UnixSocket.class);

    private AFUNIXSocket socket = null;

    @Override
    public void connect(SocketAddress endpoint, int timeout) throws IOException {
        if (!InetSocketAddress.class.isInstance(endpoint)) {
            throw new IllegalArgumentException("Expected endpoint to be a InetSocketAddress");
        }

        InetSocketAddress inetSocketAddress = (InetSocketAddress) endpoint;
        InetAddress address = inetSocketAddress.getAddress();
        String socketPath = decodeHostname(address);
        log.debug("connect via '{}'...", socketPath);

        File socketFile = new File(socketPath);

        socket = AFUNIXSocket.newInstance();

        int socketTimeout = timeout;
        if (timeout < 0) {
            socketTimeout = 0;
        }
        socket.connect(new AFUNIXSocketAddress(socketFile), socketTimeout);
        socket.setSoTimeout(socketTimeout);
    }

    @Override
    public InputStream getInputStream() throws IOException {
        if (socket == null) {
            throw new SocketException("Socket is not initialized");
        }
        return socket.getInputStream();
    }

    @Override
    public OutputStream getOutputStream() throws IOException {
        if (socket == null) {
            throw new SocketException("Socket is not initialized");
        }
        return socket.getOutputStream();
    }

    @Override
    public void bind(SocketAddress bindpoint) throws IOException {
        if (socket == null) {
            throw new SocketException("Socket is not initialized");
        }
        socket.bind(bindpoint);
    }

    @Override
    public boolean isConnected() {
        return socket != null &&  socket.isConnected();
    }

    @Override
    public boolean isClosed() {
        return socket != null && socket.isClosed();
    }

    @Override
    public void close() throws IOException {
        if (socket != null) {
            synchronized (this) {
                socket.close();
            }
        }
    }
}

File: src/main/kotlin/de/gesellix/docker/client/filesocket/UnixSocket.kt
package de.gesellix.docker.client.filesocket

import mu.KotlinLogging
import org.newsclub.net.unix.AFUNIXSocket
import org.newsclub.net.unix.AFUNIXSocketAddress
import java.io.File
import java.io.InputStream
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.SocketAddress

private val log = KotlinLogging.logger {}

class UnixSocket : FileSocket() {

    var socket: AFUNIXSocket? = null

    @Override
    override fun connect(endpoint: SocketAddress, timeout: Int) {
        if (endpoint !is InetSocketAddress) {
            throw IllegalArgumentException("expected endpoint to be a InetSocketAddress")
        }

        val address = endpoint.address
        val socketPath = decodeHostname(address)

        log.debug("connect via '$socketPath'...")
        val socketFile = File(socketPath)

        socket = AFUNIXSocket.newInstance()

        var socketTimeout = timeout
        if (timeout < 0) {
            socketTimeout = 0
        }
        socket?.connect(AFUNIXSocketAddress(socketFile), socketTimeout)
        socket?.soTimeout = socketTimeout
    }

    @Override
    override fun getInputStream(): InputStream? {
        return socket?.inputStream
    }

    @Override
    override fun getOutputStream(): OutputStream? {
        return socket?.outputStream
    }

    @Override
    override fun bind(bindpoint: SocketAddress) {
        socket?.bind(bindpoint)
    }

    @Override
    override fun isConnected(): Boolean {
        return if (socket != null) socket!!.isConnected else false
    }

    @Override
    override fun close() {
        synchronized(this) {
            socket?.close()
        }
    }
}
